### PR TITLE
magento/partners-magento2b2b#370: determine mftf coverage in the Company, CompanyCredit, CompanyPayment, CompanyShipping, & ConfigurableSharedCatalog Modules

### LINKED ISSUE #30156: URL Rewrites not loading - Cannot read property 'getData' of undefined
### Preconditions (*)

1. dev-2.4-develop (instance given by magento deployment bot)

### Steps to reproduce (*)

1. Open Admin Backend.
2. Go to: Marketing -> SEO & Search: Url Rewrites

### Expected result (*)

1. should load

### Actual result (*)

It just loads forever and nothing happens.

I get this in the browser console:

```
provider.js:121 Uncaught TypeError: Cannot read property 'getData' of undefined
    at UiClass.reload (provider.js:121)
    at invoke (resolver.js:90)
    at Array.forEach (<anonymous>)
    at resolve (resolver.js:100)
    at resolver.js:145
    at later (underscore.js:821)
```

---
Please provide [Severity](https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [x] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

### PR DESCRIPTION
**Pull request by dhaecker**
### Related Pull Requests
https://github.com/magento/partners-magento2ee/pull/420
https://github.com/magento/partners-magento2b2b/pull/496

### Fixed Issues (if relevant)

1. magento/partners-magento2b2b#370: [Determine Coverage]: Add Meaningful Titles and Descriptions For All MFTF Tests in the Company, CompanyCredit, CompanyPayment, CompanyShipping, & ConfigurableSharedCatalog Modules
1. https://jira.corp.magento.com/browse/MC-39333:  Grids Frequently (but randomly) Fail to Load & Throw getData Undefined Error in Console

### Manual testing scenarios (*)

1. ...
2. ...

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### Approval Ticket
https://jira.corp.magento.com/browse/MC-39488

### CODE REVIEW

File: app/code/Magento/Checkout/Test/Mftf/ActionGroup/StorefrontClickOrderLinkFromCheckoutSuccessPageActionGroup.xml [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+<actionGroups xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+              xsi:noNamespaceSchemaLocation="urn:magento:mftf:Test/etc/actionGroupSchema.xsd">
+    <actionGroup name="StorefrontClickOrderLinkFromCheckoutSuccessPageActionGroup">
+        <annotations>
+            <description>Clicks the order number link from the checkout success page</description>
+        </annotations>
+        <click selector="{{CheckoutSuccessMainSection.orderLink}}" stepKey="clickOrderLink"/>
+        <waitForPageLoad stepKey="waitForPageLoad"/>
+        <waitForText selector="{{StorefrontCustomerAccountMainSection.pageTitle}}" userInput="Order # " stepKey="waitForPageTitle"/>
```
Comment (lbajsarowicz) [inline app/code/Magento/Checkout/Test/Mftf/ActionGroup/StorefrontClickOrderLinkFromCheckoutSuccessPageActionGroup.xml]:
```suggestion
        <waitForText selector="{{StorefrontCustomerAccountMainSection.pageTitle}}" userInput="Order #" stepKey="waitForPageTitle"/>
```

File: app/code/Magento/Customer/Test/Mftf/ActionGroup/AdminSetCustomerActiveViaGridActionGroup.xml [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+        <arguments>
+            <argument name="customerEmail" type="string" defaultValue="{{Simple_US_CA_Customer.email}}"/>
+        </arguments>
+        <waitForElementVisible selector="{{AdminCustomerGridMainActionsSection.customerCheckbox(customerEmail)}}" stepKey="waitForCustomerRow"/>
+        <click selector="{{AdminCustomerGridMainActionsSection.customerCheckbox(customerEmail)}}" stepKey="selectCustomer"/>
+        <click selector="{{AdminCustomerGridMainActionsSection.actions}}" stepKey="clickActions"/>
+        <waitForElementVisible selector="{{AdminCustomerGridMainActionsSection.setActive}}" stepKey="waitForDropDownOpen"/>
+        <click selector="{{AdminCustomerGridMainActionsSection.setActive}}" stepKey="clickSetActive"/>
+        <waitForPageLoad stepKey="waitForLoad"/>
+        <waitForText selector="{{AdminMessagesSection.success}}" userInput="A total of 1 record(s) were updated." stepKey="seeSuccessMessage"/>
```
Comment (lbajsarowicz) [inline app/code/Magento/Customer/Test/Mftf/ActionGroup/AdminSetCustomerActiveViaGridActionGroup.xml:24]: Could you use existing `Assert` Action Group in Test, as in some cases we might like to assert that the action fails?

File: app/code/Magento/Customer/Test/Mftf/ActionGroup/StorefrontGoToCustomerAddressesPageActionGroup.xml [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+  */
+-->
+<actionGroups xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+              xsi:noNamespaceSchemaLocation="urn:magento:mftf:Test/etc/actionGroupSchema.xsd">
+    <actionGroup name="StorefrontGoToCustomerAddressesPageActionGroup">
+        <annotations>
+            <description>Goes to the storefront customer addresses page</description>
+        </annotations>
+        <amOnPage url="{{StorefrontCustomerAddressesPage.url}}" stepKey="goToAddressPage"/>
+        <waitForPageLoad stepKey="waitForPageLoad"/>
```
Comment (lbajsarowicz) [inline app/code/Magento/Customer/Test/Mftf/ActionGroup/StorefrontGoToCustomerAddressesPageActionGroup.xml]: `<amOnPage` already includes `waitForPageLoad` in its implementation, you don't need to add it manually.

**Review by lbajsarowicz — CHANGES_REQUESTED** [review]
🟡  I have some minor request for you, regarding Action Group reusability. Really appreciate your input!

File: app/code/Magento/Customer/Test/Mftf/ActionGroup/AdminSetCustomerActiveViaGridActionGroup.xml (hunk 2 above)
Comment (dhaecker) [reply app/code/Magento/Customer/Test/Mftf/ActionGroup/AdminSetCustomerActiveViaGridActionGroup.xml:24]: @lbajsarowicz are you suggesting that i should remove the `waitForText` line from this actiongroup & use an assert actiongroup after i reference `AdminSetCustomerActiveViaGridActionGroup` in the test?

my assumption was that this actiongroup i wrote could be for the positive scenario & if someone needs to specifically test the negative case, then they could extend this actiongroup & merge with the `waitForText` action 🤷 

what are your thoughts between the two approaches?

Comment (lbajsarowicz) [reply app/code/Magento/Customer/Test/Mftf/ActionGroup/AdminSetCustomerActiveViaGridActionGroup.xml:24]: You can rename the Action Group then to include state assertion `Successfully`.

File: app/code/Magento/Checkout/Test/Mftf/ActionGroup/StorefrontClickOrderLinkFromCheckoutSuccessPageActionGroup.xml (hunk 1 above)
Comment (dhaecker) [reply app/code/Magento/Checkout/Test/Mftf/ActionGroup/StorefrontClickOrderLinkFromCheckoutSuccessPageActionGroup.xml]: done

File: app/code/Magento/Customer/Test/Mftf/ActionGroup/AdminSetCustomerActiveViaGridActionGroup.xml (hunk 2 above)
Comment (dhaecker) [reply app/code/Magento/Customer/Test/Mftf/ActionGroup/AdminSetCustomerActiveViaGridActionGroup.xml:24]: done

File: app/code/Magento/Customer/Test/Mftf/ActionGroup/StorefrontGoToCustomerAddressesPageActionGroup.xml (hunk 3 above)
Comment (dhaecker) [reply app/code/Magento/Customer/Test/Mftf/ActionGroup/StorefrontGoToCustomerAddressesPageActionGroup.xml]: done

**Comment by lbajsarowicz** [comment]
@magento run all tests

File: app/code/Magento/Customer/Test/Mftf/ActionGroup/StorefrontCustomStoreCustomerLogoutActionGroup.xml [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+-->
+
+<actionGroups xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+              xsi:noNamespaceSchemaLocation="urn:magento:mftf:Test/etc/actionGroupSchema.xsd">
+    <actionGroup name="StorefrontCustomStoreCustomerLogoutActionGroup" extends="StorefrontCustomerLogoutActionGroup">
+        <annotations>
+            <description>Goes to the storefront Customer Logout page for a custom store. Must have Add Store Code To Urls enabled.</description>
+        </annotations>
+        <arguments>
+            <argument name="storeCode"/>
```
Comment (danmooney2) [inline app/code/Magento/Customer/Test/Mftf/ActionGroup/StorefrontCustomStoreCustomerLogoutActionGroup.xml]: type="string" would help?

File: app/code/Magento/Customer/Test/Mftf/ActionGroup/StorefrontCustomStoreLoginActionGroup.xml [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+-->
+
+<actionGroups xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+              xsi:noNamespaceSchemaLocation="urn:magento:mftf:Test/etc/actionGroupSchema.xsd">
+    <actionGroup name="StorefrontCustomStoreLoginActionGroup" extends="LoginToStorefrontActionGroup">
+        <annotations>
+            <description>Goes to the storefront Customer Sign In page for a custom store. Logs in using the provided Customer. Must have Add Store Code To Urls enabled</description>
+        </annotations>
+        <arguments>
+            <argument name="storeCode"/>
```
Comment (danmooney2) [inline app/code/Magento/Customer/Test/Mftf/ActionGroup/StorefrontCustomStoreLoginActionGroup.xml]: type="string" would help?

File: app/code/Magento/Customer/Test/Mftf/ActionGroup/StorefrontCustomStoreNavigateToCustomerOrdersHistoryPageActionGroup.xml [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+  */
+-->
+<actionGroups xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+              xsi:noNamespaceSchemaLocation="urn:magento:mftf:Test/etc/actionGroupSchema.xsd">
+    <actionGroup name="StorefrontCustomStoreNavigateToCustomerOrdersHistoryPageActionGroup" extends="StorefrontNavigateToCustomerOrdersHistoryPageActionGroup">
+        <annotations>
+            <description>Goes to the storefront Customer Order History page for a custom store. Must have Add Store Code To Urls enabled</description>
+        </annotations>
+        <arguments>
+            <argument name="storeCode"/>
```
Comment (danmooney2) [inline app/code/Magento/Customer/Test/Mftf/ActionGroup/StorefrontCustomStoreNavigateToCustomerOrdersHistoryPageActionGroup.xml]: type="string" would help?

**Review by danmooney2 — CHANGES_REQUESTED** [review]

**Comment by gabrieldagama** [comment]
@magento create issue

**Comment by dhaecker** [comment]
@gabrieldagama issue already exists in partners-b2b. this PR is simply supporting the b2b PR & issue

File: app/code/Magento/Customer/Test/Mftf/ActionGroup/StorefrontCustomStoreCustomerLogoutActionGroup.xml (hunk 4 above)
Comment (dhaecker) [reply app/code/Magento/Customer/Test/Mftf/ActionGroup/StorefrontCustomStoreCustomerLogoutActionGroup.xml]: yes good call. will update

**Comment by dhaecker** [comment]
@magento run Functional Tests CE

**Review by danmooney2 — APPROVED** [review]

**Review by engcom-Delta — CHANGES_REQUESTED** [review]
Hi @dhaecker I was able to reproduce issue in https://jira.corp.magento.com/browse/MC-39333 . With PR changes this issue is fixed, but i got another error in dev console:
```
Uncaught TypeError: Cannot read property 'offsetLeft' of undefined
    at UiClass.checkTableWidth (toolbar.js:363)
    at UiClass.run (toolbar.js:237)
    at Object.<anonymous> (jquery.js:3417)
    at fire (jquery.js:3238)
    at Object.fireWith [as resolveWith] (jquery.js:3368)
    at Object.deferred.<computed> [as resolve] (jquery.js:3467)
    at invoke (resolver.js:90)
    at Array.forEach (<anonymous>)
    at resolve (resolver.js:100)
    at resolver.js:145
```
**Manual testing scenario:**
1. Set env to development mode
2. Delete Magento_Ui/js/grid/data-storage.js from pub/static (in my case I used `rm -rf pub/static/*`)
3. Add following snippet to pub/static.php above bootstrap instantiation:
```
if (strpos($_REQUEST['resource'], 'Magento_Ui/js/grid/data-storage.js') !== false) {
    sleep(8);
}
```
4. Go to any page where UI component grid loads with console open. Ensure Magento_Ui/js/grid/data-storage.js takes significant time (>8 seconds) to return a response

**Result:**
:x:  Note error in console `Cannot read property 'offsetLeft' of undefined`
#30957Pr


Could you take a look?

**Comment by dhaecker** [comment]
@magento run Functional Tests CE

**Comment by dhaecker** [comment]
@magento run all tests
